private/pkg/manifest/storage.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Manifests are a list of paths and their hash digests, canonically ordered by
+// path in increasing lexographical order. Manifests are encoded as:
+//
+//	<hash type>:<digest>[SP][SP]<path>[LF]
+//
+// "shake256" is the only supported hash type. The digest is 64 bytes of hex
+// encoded output of SHAKE256. See golang.org/x/crypto/sha3 and FIPS 202 for
+// details on the SHAKE hash.
+package manifest
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"github.com/bufbuild/buf/private/pkg/storage/storagemem"
+	"go.uber.org/multierr"
+)
+
+// NewFromBucket creates a manifest from a storage bucket, with all its digests
+// in DigestTypeShake256.
+func NewFromBucket(
+	ctx context.Context,
+	bucket storage.ReadBucket,
+) (*Manifest, error) {
+	m := New()
+	digester, err := NewDigester(DigestTypeShake256)
+	if err != nil {
+		return nil, err
+	}
+	if walkErr := bucket.Walk(ctx, "", func(info storage.ObjectInfo) (retErr error) {
+		path := info.Path()
+		obj, err := bucket.Get(ctx, path)
+		if err != nil {
+			return err
+		}
+		defer func() { retErr = multierr.Append(retErr, obj.Close()) }()
+		digest, err := digester.Digest(obj)
+		if err != nil {
+			return err
+		}
+		if err := m.AddEntry(path, *digest); err != nil {
+			return err
+		}
+		return nil
+	}); walkErr != nil {
+		return nil, walkErr
+	}
+	return m, nil
+}
+
+// ToBucket takes a map of digest to files' contents, and builds a storagemem
+// bucket from the manifest.
+func (m *Manifest) ToBucket(digestToFiles map[string][]byte) (storage.ReadBucket, error) {
```

Comment:
Uncertain about what to receive here. For this PR I'm receiving:

```
{
  "shake256:the-digest": "the actual file content",
}
```

But we could receive:

```
{
  "shake256:the-digest": io.Reader,
}
```

Although in most scenarios I imagine the caller already has the content in memory (like from the backend, in blob.Content), so no need for the indirection?

Also, not receiving:

```
{
  "file-path": "content or reader",
}
```

To avoid content duplication.

Another thing I considered, but not sure if it's a good idea to put that much logic into the Manifest, would be making sure the content matches the received content, by rehashing it. That way we could even just receive an array of de-duplicated contents.

Reply:
`digestToFiles map[string][]byte` ties pretty hard to a string representation of a digest and content being in memory. Even `[]byte` is changed to an io.Reader, it's still a forced string representation.

`NewDigestFromBlobHash` makes a nice Digest type, so what if we make a Blob abstraction from a proto Blob? Something that has a Digest and can be an io.ReadCloser.

Warning: hastily thrown together code below! Figure out where to put it and maybe even make it work!

Let's define what a blob is and make a in-memory blob implementation.

```go
type Blob interface {
  Digest() *Digest
  Open(ctx context.Context) (io.ReadCloser, error)
}

type MemoryBlob struct {
  digest *Digest
  content []byte
}

var _ Blob = (*InMemoryBlob)(nil)

func NewMemoryBlob(digest *Digest, content []byte) *MemoryBlob {
  return &MemoryBlob{digest: digest, content: content}
}

func (memblob *MemoryBlob) Digest() *Digest {
  return memblob.digest
}

func (memblob *MemoryBlob) Open(_ context.Context) io.ReadCloser {
  return bytes.NewReader(memblob.content), nil
}
```

Then if we tie the open'd path to an io.ReadCloser:

```go
type storageCanBeSilly struct {
   path string
   file io.ReadCloser
}

func (silly *storageCanBeSilly) Path() string { return silly.path}
func (silly *storageCanBeSilly) ExternalPath() string { return silly.path}
func (silly *storageCanBeSilly) Read(p []bytes) (int, error) { return silly.file.Read(p) }
func (silly *storageCanBeSilly) Close() error { return silly.file.Close() }
```

Now we can implement a storage.ReadBucket by tying a manifest and an array of blobs together.

```go
type blobBucket struct {
  bag map[string]Blob
}

func NewBlobBucket(manifest *Manifest, blobs []Blob) storage.ReadBucket {
  bb := &blobBucket{bag: make(map[string]Blob])}
  for _, blob := range blobs {
    bb.bag[blob.Digest().String()] = blob
  }
  return bb
}

func (bb *blobBucket) Get(ctx context.Context, path string) (ReadObjectCloser, error) {
  digest, ok := bb.manifest.DigestFor(path)
  if !ok {
    return nil, errors.New("not found in manifest")
  }
  blob, ok := bb.bag[digest.String()]
  if !ok {
    return nil, errors.New("not found in blob bag")
  }
  file, err := blob.Open(ctx)
  if err != nil {
    return nil, err
  }
  return &storageCanBeSilly{path: path, file: file}, nil
}
```

This works for in-memory *and* can support on-disk or in GCS blobs. Those just need a different implementation if you want to be memory efficient.

Reply:
Thanks, that's a great suggestion. Pushed it!